src/Runner/Parallel/Process.php

```diff
@@ -0,0 +1,260 @@
+        RunnerConfig $runnerConfig,
+        ProcessIdentifier $identifier,
+        int $serverPort
+    ): self {
+        $input = self::getArgvInput();
+        $phpBinary = (new PhpExecutableFinder())->find(false);
+
+        if (false === $phpBinary) {
+            throw new ParallelisationException('Cannot find PHP executable.');
+        }
+
+        $commandArgs = [
+            $phpBinary,
+            escapeshellarg($_SERVER['argv'][0]),
+            'worker',
+            '--port',
+            (string) $serverPort,
+            '--identifier',
+            escapeshellarg((string) $identifier),
+        ];
+
+        if ($runnerConfig->isDryRun()) {
+            $commandArgs[] = '--dry-run';
+        }
+
+        if (filter_var($input->getOption('diff'), FILTER_VALIDATE_BOOLEAN)) {
+            $commandArgs[] = '--diff';
+        }
+
+        foreach (['allow-risky', 'config', 'rules', 'using-cache', 'cache-file'] as $option) {
+            $optionValue = $input->getOption($option);
+
+            if (null !== $optionValue) {
+                $commandArgs[] = "--{$option}";
+                $commandArgs[] = escapeshellarg($optionValue);
+            }
+        }
+
+        return new self(
+            implode(' ', $commandArgs),
+            $loop,
+            $runnerConfig->getParallelConfig()->getProcessTimeout()
+        );
+    }
+
+    /**
+     * @param callable(mixed[] $json): void                  $onData  callback to be called when data is received from the parallelisation operator
+     * @param callable(\Throwable $exception): void          $onError callback to be called when an exception occurs
+     * @param callable(?int $exitCode, string $output): void $onExit  callback to be called when the process exits
+     */
+    public function start(callable $onData, callable $onError, callable $onExit): void
+    {
+        $this->stdOut = tmpfile();
+        if (false === $this->stdOut) {
+            throw new ParallelisationException('Failed creating temp file for stdOut.');
+        }
+
+        $this->stdErr = tmpfile();
+        if (false === $this->stdErr) {
+            throw new ParallelisationException('Failed creating temp file for stdErr.');
+        }
+
+        $this->onData = $onData;
+        $this->onError = $onError;
+
+        $this->process = new ReactProcess($this->command, null, null, [
+            1 => $this->stdOut,
+            2 => $this->stdErr,
+        ]);
+        $this->process->start($this->loop);
+        $this->process->on('exit', function ($exitCode) use ($onExit): void {
+            $this->cancelTimer();
+
+            $output = '';
+            rewind($this->stdOut);
+            $stdOut = stream_get_contents($this->stdOut);
+            if (\is_string($stdOut)) {
+                $output .= $stdOut;
+            }
+
+            rewind($this->stdErr);
+            $stdErr = stream_get_contents($this->stdErr);
+            if (\is_string($stdErr)) {
+                $output .= $stdErr;
```

Comment:
Is it possible that appending all stderr after stdout would give weird results because both streams are actually written in parallel?

Reply:
The worker's output currently is completely dropped, I'll cover it when I get to "error handling" todo entry. Anyway, the worker process' output is used only [here](https://github.com/Wirone/PHP-CS-Fixer/blob/4db0aa01879cd59d99851cd4efe1bf88dd3c2022/src/Runner/Runner.php#L279-L287), when process exits. [In PHPStan](https://github.com/phpstan/phpstan-src/blob/6320e1d2fd9d2bc23dc528e3296ab15bdee3522c/src/Parallel/ParallelAnalyser.php#L234-L269) it just scans the output looking for known errors and propagates them to the main process. I don't know yet how I'll approach it, but answering your initial question: no, it does not affect the main process because it's totally separated 🙂.